### ISSUE TITLE
enhancement Identity Header

### ISSUE THREAD

**Issue by gilles0606** [irontec/sngrep#463]
hello,

it would be nice to have at least the decoding of the JWT in the INVITE message, 
or by selecting the Identity Header in INVITE Display ?

Thanks

**Comment by altzone** [comment]
Yes, that would be great, or even initially to see if the identity field is present and add a column to know if the call is signed.

**Comment by Kaian** [comment]
Hi @ooglek,

Yes, feel free to send me the PCAP via email (or attaching a zip to this issue)

sngrep doesn’t mind additional headers beyond the basic ones (e.g., Call-ID, From, To), so I’m not sure why the Identity header would behave any differently.

As for decoding JWTs, that would require adding an external library (and making it optional) to parse the header contents.

There was a stalled branch, _glib-2_, which introduced some configuration options to create dynamic columns by parsing the initial dialog messages with regex. However, it's a buggy version, so I wouldn’t recommend using it in production environments.

The current version of sngrep is in bugfix-only mode, so issues like skipped messages or unexpected crashes are definitely worth reporting and fixing.

Best regards,
Kaian

**Comment by ooglek** [comment]
@Kaian I think the issue is NOT Identity headers now. I think the issue is that when packets are loaded from the PCAP that are fragmented, either the PCAP file does not contain the fragments, or it does and `sngrep` does not put the fragments together. 

This may be related to #499 where INVITEs are fragmented and then do not show up in sngrep. 

However, I am not yet fully confident that my PCAP file contains both the fragmented INVITE packet AND the rest of the next packet. I'll continue to do some investigation and let you know.

**Comment by ooglek** [comment]
### I revoke my issue.

tl;dr - My packet capture was not capturing UDP packet fragments. When listening to the ethernet device (`-d eth1`), sngrep did the correct thing and re-assembled packets correctly.


How I got here:

I was using a long-running `tcpdump` command to capture packets. However, before `Identity:` headers, most of the `INVITE`s would come in under the 1500 byte IP limit.

With the additional data, `INVITE` packets now exceed 1500 bytes and are fragmented, but my command was not capturing that.

This is the new and improved `tcpdump` command that captures both UDP 5060 SIP Traffic, AND UDP Fragments. `sngrep` correctly re-assembles them.

**NOTE:** Fragments do NOT contain the UDP Port, so the 2nd pcap filter saves ALL UDP Fragments, even if they aren't on UDP Port 5060. If there's a way to keep just the 5060-related fragments, I'm open to learning.

```
/usr/sbin/tcpdump -qpni eth0 -s 65535 -C 100 -W 90 -Z root -w /var/log/sip.pcap (udp port 5060 and not udp[8:4] = 0x4f505449) or (udp and ip[6:2] & 0x3fff != 0x0000)
```

1. Includes all Port 5060 UDP Traffic (`udp port 5060`)
2. UNLESS it is an OPTIONS command (`not udp[8:4] = 0x4f505449`)
3. OR if it is UDP AND (either More Fragments bit is set OR the fragment offset is non-zero) (`udp and ip[6:2] & 0x3fff != 0x0000`)

The last item came from the [tcpdump mailing lists on 'fragmented ip packets' from 2010](https://seclists.org/tcpdump/2010/q4/94).

> Here's a filter that will match all fragments:
> "ip[6:2] & 0x3fff != 0x0000"
> 
> Essentially, it is a combination of the following two filters:
> "ip[6:2] & 0x2000 == 0x2000"
> "ip[6:2] & 0x1fff != 0x0000"
> 
> The first filter checks if the "More Fragments" bit is set, so it will match all
> fragments except for the last one.  The second filter checks if the fragment
> offset is non-zero, so it will match all fragments except for the first one. 
> Used together, you will get all IP fragments.
